### ISSUE TITLE
You Tube: Example of integrating candidate model with the ranking model

### ISSUE THREAD

**Issue by Sade-KD** [massquantity/LibRecommender#14]
From the paper, the you tube model has two major components candidate (i.e you tube matcher script) and the ranking model (i.e. youtube ranking script). 

![CModel](https://user-images.githubusercontent.com/42578613/96132616-e6b27500-0efa-11eb-9081-eaf5690a1ece.PNG)

![rModel](https://user-images.githubusercontent.com/42578613/96132802-0b0e5180-0efb-11eb-8e3d-8cf51df45879.PNG)

The features passed into the ranking component and candidate generation component are different and are selected based on their purpose. 

Can you kindly elaborate more this aspect and to please provide a generic example of the you-tube model as a whole with the ranking model and candidate generation model using the movie lens dataset?

**Comment by massquantity** [comment]
Based on the paper, the candidate generation model only used user features, because its purpose was to retrieve relevant items from large item pool at high speed. After the candidate generation model successfully retrieved some, say 1000 items, the  ranking component would be used to rank these items and select maybe top 10 items to recommend. The ranking component  generally would use a more complex model and more features for high accuracy.

In LibRecommender, however, the two models are treated separately. In other words, you can only use one of them for training and recommending, but they can't be combined together like the procedure described above.

**Comment by massquantity** [comment]
I feel you know how to use `YouTubeRanking` model. As for `YouTubeMatch` model, here is an example:

```python
import pandas as pd
from libreco.data import DatasetFeat
from libreco.data import (split_by_ratio_chrono,)
from libreco.algorithms import YouTubeMatch

data = pd.read_csv(...)
train, test = split_by_ratio_chrono(data, test_size=0.2)

sparse_col = ["sex", "occupation"]
dense_col = ["age"]
user_col = ["sex", "age", "occupation"]
item_col = []    # empty item feature
train_data, data_info = DatasetFeat.build_trainset(train, user_col, item_col,
                                                   sparse_col, dense_col,
                                                   shuffle=False)
test_data = DatasetFeat.build_testset(test, sparse_col, dense_col, shuffle=False)
print(data_info)

ytb_match = YouTubeMatch("ranking", data_info, embed_size=16, n_epochs=100,
                         lr=1e-4, lr_decay=False, reg=None, batch_size=2048,
                         num_neg=1, use_bn=False, dropout_rate=None,
                         loss_type="nce", hidden_units="128,64,32",
                         tf_sess_config=None)
ytb_match.fit(train_data, verbose=2, shuffle=True, eval_data=None,)
print("prediction: ", ytb_match.predict(1, 2))
print("recommendation: ", ytb_match.recommend_user(1, 7))
```

One thing worth mentioning, this model can't use item features, as revealed by the original paper. If you want to evaluate model during training, you need to sample negative items first:

```python
data = pd.read_csv(...)
train, test = split_by_ratio_chrono(data, test_size=0.2)

sparse_col = ["sex", "occupation"]
dense_col = ["age"]
user_col = ["sex", "age", "occupation"]
item_col = []
train_data, data_info = DatasetFeat.build_trainset(train, user_col, item_col,
                                                   sparse_col, dense_col,
                                                   shuffle=False)
test_data = DatasetFeat.build_testset(test, sparse_col, dense_col, shuffle=False)
print(data_info)

# sampling negative  procedure
data["label"] = 1
train_data.build_negative_samples(data_info, item_gen_mode="random")
test_data.build_negative_samples(data_info, item_gen_mode="random")

ytb_match = YouTubeMatch("ranking", data_info, embed_size=16, n_epochs=100,
                         lr=1e-4, lr_decay=False, reg=None, batch_size=2048,
                         num_neg=1, use_bn=False, dropout_rate=None,
                         loss_type="nce", hidden_units="128,64,32",
                         tf_sess_config=None)
ytb_match.fit(train_data, verbose=2, shuffle=True, eval_data=test_data,  # add eval data
              metrics=["loss", "balanced_accuracy", "roc_auc", "pr_auc",
                       "precision", "recall", "map", "ndcg"])
```

**Comment by Sade-KD** [comment]
> Based on the paper, the candidate generation model only used user features, because its purpose was to retrieve relevant items from large item pool at high speed. After the candidate generation model successfully retrieved some, say 1000 items, the ranking component would be used to rank these items and select maybe top 10 items to recommend. The ranking component generally would use a more complex model and more features for high accuracy.
> 
> In LibRecommender, however, the two models are treated separately. In other words, you can only use one of them for training and recommending, but they can't be combined together like the procedure described above.

I am glad I asked this question, since I was trying to link back to paper and a few things wont aligning and your explanation confirmed and clarified it.

My last question is on the details of the training and hyperparameter tuning the model. However, it might be best if I open it up as new issue since its a key component.
